### DISCUSSION TITLE
Chatroom example problems.

### DISCUSSION THREAD

**Discussion by murodjon001** [django/channels#1987] [Help!] [answered]
I want to create a chatroom between users using channels in my django project! I wrote all the routing, consumer and other python files using the documentation but I got an error. Please help me to fix this error!here is the error i am getting:TypeError: object ChatRoomConsumer can't be used in 'await' expression WebSocket DISCONNECT /ws/chat/asma/ [127.0.0.1:56540]

from django.urls import re_path

from . import consumers

websocket_urlpatterns = [
    re_path(r'ws/chat/(?P<room_name>\w+)/$', consumers.ChatRoomConsumer),
]

**Comment by murodjon001** [comment]
# this consumers.py  
import json
from asgiref.sync import async_to_sync,sync_to_async

from channels.generic.websocket import AsyncWebsocketConsumer

                
class ChatRoomConsumer(AsyncWebsocketConsumer):
    async def connect(self):
        self.room_name = self.scope['url_route']['kwargs']['room_name']
        self.room_group_name = 'chat_%s' % self.room_name

        await self.channel_layer.group_add(
            self.room_group_name,
            self.channel_name
        )

        

        await self.channel_layer.group_send(
            self.room_group_name,
            {
                'type': 'tester_message',
                'tester': 'helloWorld',
            }

        )
        print('connect ishladi')
        self.accept()
        
    async def tester_message(self,event):
        tester = event['tester']

        await self.send(text_data=json.dumps({
            'tester': tester,
        }))

        

    async def disconnect(self, close_code):
        # return await super().disconnect(self, close_code):  bu kod avtomatik chiqdi ishlatib ko'rish kerak
        await self.channel_layer.group_discard(
            self.room.group_name,
            self.channel_name,
            print('disconnect ishladi')
        )

**Comment by murodjon001** [comment]
WSGI_APPLICATION = "hosptalsam.wsgi.application"
ASGI_APPLICATION = "hosptalsam.asgi.application"


ASGI_APPLICATION = "hosptalsam.routing.application"
CHANNEL_LAYERS = {
    "default": {
        "BACKEND": "channels_redis.core.RedisChannelLayer",
        "CONFIG": {
            "hosts": [("127.0.0.1", 6379)],
        },
    },
}

**Comment by murodjon001** [comment]
# this hosptalsam/routing.url (core)
from message import routing
from django.core.asgi import get_asgi_application
from django.urls import re_path

# Initialize Django ASGI application early to ensure the AppRegistry
# is populated before importing code that may import ORM models.
django_asgi_app = get_asgi_application()

from channels.routing import ProtocolTypeRouter, URLRouter
from channels.auth import AuthMiddlewareStack
from channels.security.websocket import AllowedHostsOriginValidator

application = ProtocolTypeRouter({
    "http": django_asgi_app,
    "websocket": AllowedHostsOriginValidator(
        AuthMiddlewareStack(
            URLRouter(
                routing.websocket_urlpatterns
            )
        )
    ),
})

**Comment by murodjon001** [comment]
asgiref==3.6.0
async-timeout==4.0.2
attrs==22.2.0
autobahn==23.1.2
Automat==22.10.0
backports.zoneinfo==0.2.1
cffi==1.15.1
channels==4.0.0
channels-redis==4.0.0
constantly==15.1.0
cryptography==39.0.2
daphne==4.0.0
Django==4.1.7
django-jazzmin==2.6.0
hyperlink==21.0.0
idna==3.4
incremental==22.10.0
msgpack==1.0.5
Pillow==9.4.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pyOpenSSL==23.0.0
redis==4.5.1
service-identity==21.1.0
six==1.16.0
sqlparse==0.4.3
Twisted==22.10.0
txaio==23.1.1
typing-extensions==4.5.0
zope.interface==5.5.2

**Comment by murodjon001** [comment]
WebSocket HANDSHAKING /ws/chat/asma/ [127.0.0.1:56540]
Exception inside application: object ChatRoomConsumer can't be used in 'await' expression
Traceback (most recent call last):
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/django/contrib/staticfiles/handlers.py", line 101, in __call__
    return await self.application(scope, receive, send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/routing.py", line 62, in __call__
    return await application(scope, receive, send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/security/websocket.py", line 37, in __call__
    return await self.application(scope, receive, send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/sessions.py", line 47, in __call__
    return await self.inner(dict(scope, cookies=cookies), receive, send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/sessions.py", line 263, in __call__
    return await self.inner(wrapper.scope, receive, wrapper.send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/auth.py", line 185, in __call__
    return await super().__call__(scope, receive, send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/middleware.py", line 24, in __call__
    return await self.inner(scope, receive, send)
  File "/home/murodjon/Desktop/texnikum_web/myenv/lib/python3.8/site-packages/channels/routing.py", line 116, in __call__
    return await application(
TypeError: object ChatRoomConsumer can't be used in 'await' expression
WebSocket DISCONNECT /ws/chat/asma/ [127.0.0.1:56540]

**Comment by carltongibson** [comment, marked as answer]
> `re_path(r'ws/chat/(?P<room_name>\w+)/$', consumers.ChatRoomConsumer),`

> `TypeError: object ChatRoomConsumer can't be used in 'await' expression`

You're missing an `.as_asgi()`: 

```
re_path(r'ws/chat/(?P<room_name>\w+)/$', consumers.ChatRoomConsumer.as_asgi()),
```

**Reply by murodjon001** [reply]
Thank you very much for the answer! My code worked!
